File: customer-api-spa/types/planner/PlannerExcursion.ts
import { Uuid } from '../basic/Uuid'
import { ExcursionFilterParams, makeExcursionFilterParams } from './ExcursionFilterParams'
import { fromQueryGroup, QueryClientGroup } from '../bookings/QueryClientGroup'
import { QueryGroup } from '../bookings/QueryGroup'
import { v4 as uuid } from 'uuid'
import { PlannerExcursionDate } from './PlannerExcursionDates'
import { DateString } from '../basic/DateString'
import { PlannerExcursionVariantDate } from './PlannerExcursionVariantDate'
import { PlannerExcursionItemId } from './PlannerExcursionItem'

export interface PlannerExcursion {
  id: Uuid;
  queryGroup: QueryClientGroup;
  dates: DateString[];
  items: PlannerExcursionVariantDate[];
  selected: PlannerExcursionItemId[];
  variantCount: number;
  excursionDates: PlannerExcursionDate[];
  filter: ExcursionFilterParams;
  maxItems: number;
}

export const makePlannerExcursion = (queryGroup: QueryGroup): PlannerExcursion => {
  const cg = fromQueryGroup(queryGroup)
  return {
    id: uuid(),
    queryGroup: cg,
    dates: [],
    items: [],
    selected: [],
    variantCount: 0,
    excursionDates: [],
    filter: makeExcursionFilterParams(),
    maxItems: 10,
  }
}

File: orders-protocol/src/main/scala/bookingtour/protocols/orders/api/PlannerExcursion.scala
package bookingtour.protocols.orders.api

import java.time.LocalDate

import bookingtour.protocols.core._
import bookingtour.protocols.orders.newTypes.{PlannerExcursionId, PlannerExcursionItemId}
import cats.instances.all._
import derevo.cats.order
import derevo.circe.{decoder, encoder}
import derevo.derive

/**
  * © Alexey Toroshchin 2020.
  */
@derive(encoder, decoder, order)
final case class PlannerExcursion(
    id: PlannerExcursionId,
    queryGroup: PlannerQueryClientGroup,
    dates: List[LocalDate],
    items: List[PlannerExcursionVariantDate],
    selected: List[PlannerExcursionItemId],
    variantCount: Int,
    excursionDates: List[PlannerExcursionDate],
    filter: PlannerExcursionFilterParams,
    maxItems: Int
)
